lib/Xud.ts

```diff
@@ -67,29 +62,36 @@ class Xud extends EventEmitter {
     this.logger.info('config loaded');
 
     try {
-      const initPromises: Promise<any>[] = [];
-      // TODO: wait for decryption of existing key or encryption of new key, config option to disable encryption
-      initPromises.push(NodeKey.load(this.config.xudir, this.config.instanceid));
-
       this.db = new DB(loggers.db, this.config.dbpath);
       await this.db.init(this.config.network, this.config.initdb);
-      this.pool = new Pool(this.config.p2p, this.config.network, loggers.p2p, this.db.models, version);
-      this.swapClientManager = new SwapClientManager(this.config, loggers, this.pool);
-      await this.swapClientManager.init(this.db.models);
 
+      const nodeKey = await NodeKey.load(this.config.xudir, this.config.instanceid);
+      this.logger.info(`Local nodePubKey is ${nodeKey.nodePubKey}`);
+
+      this.pool = new Pool({
+        nodeKey,
+        version,
+        config: this.config.p2p,
+        xuNetwork: this.config.network,
+        logger: loggers.p2p,
+        models: this.db.models,
+      });
+      this.swapClientManager = new SwapClientManager(this.config, loggers);
       this.swaps = new Swaps(loggers.swaps, this.db.models, this.pool, this.swapClientManager);
+
+      await this.swapClientManager.init(this.db.models);
```

Comment:
nit: this line could be right below the creation of `SwapClientManager` (line 79)

Reply:
So actually this was a necessary change (one that I spent a while figuring out) because `pool` needs to know about the lnd pub keys and raiden address that we get only after `swapClientManager.init` completes. `swapClientManager` notifies `swaps` whenever we get a new pubkeyor address, and `swaps` updates `pool`, and with this change that happens on initialization. If we initialize `swapClientManager` before creating `swaps`, it doesn't happen.

But in looking at this again, this seems less than ideal and fragile. What I can do is put logic in the `Swaps` constructor to update `pool` with the pub keys that we already have. That way it doesn't matter in what order we put these lines, and I can go ahead and reorder this as you've suggested. How does that sound?

Reply:
I'm not sure about adding the listeners in `Swaps` constructor. In the long run I'm beginning to think a pub-sub pattern between our services might be a better approach compared to dependency injection, but it's definitely out of scope of this PR.

I'm good with just leaving a comment right now that explains why `swapClientManager` is initalized before creating `swaps`.

Reply:
There wouldn't be any new listeners being added, just a check in the constructor to set existing lnd pubkeys and raiden addresses from `swapClientManager` to `pool`. Although I'd also be OK leaving it as is with a comment to explain.

Reply:
Up to you :).

Reply:
Putting that much code in the constructor didn't seem right - and in general I don't think calling constructors should be making changes to other objects - so I moved that logic in `init`. It works now (and they can be initialized in parallel) and it's less fragile I believe so I'll go with this if it looks good to you as well.